src/mapreduce.jl

```diff
@@ -0,0 +1,135 @@
+    v0 = startvalue(op, OT) # TODO do this better
+    mapreduce(f, op, v0, A)
+end
+function acc_mapreduce end
+function Base.mapreduce(f, op, v0, A::GPUArray, B::GPUArray, C::Number)
+    acc_mapreduce(f, op, v0, A, (B, C))
+end
+function Base.mapreduce(f, op, v0, A::GPUArray, B::GPUArray)
+    acc_mapreduce(f, op, v0, A, (B,))
+end
+function Base.mapreduce(f, op, v0, A::GPUArray)
+    acc_mapreduce(f, op, v0, A, ())
+end
+
+
+function mapreducedim_kernel(state, f, op, R::AbstractArray{T1, N}, A::AbstractArray{T, N}, slice_size, sizeA, dim) where {T1, T, N}
+    ilin = Cuint(linear_index(state))
+    ilin > length(R) && return
+    accum = zero(T1)
+    @inbounds for i = Cuint(1):slice_size
+        idx = N == dim ? (ilin, i) : (i, ilin)
+        i2d = gpu_sub2ind(sizeA, idx)
+        accum = op(accum, f(A[i2d]))
+    end
+    R[ilin] = accum
+    return
+end
+function Base._mapreducedim!(f, op, R::GPUArray, A::GPUArray)
+    sizeR = size(R)
+    if all(x-> x == 1, sizeR)
+        x = mapreduce(f, op, A)
+        copy!(R, reshape([x], sizeR))
+        return R
+    end
+    @assert count(x-> x == 1, sizeR) == (ndims(R) - 1) "Not implemented"
+    dim = findfirst(x-> x == 1, sizeR)
+    slice_size = size(A, dim)
+    gpu_call(mapreducedim_kernel, R, (f, op, R, A, Cuint(slice_size), Cuint.(size(A)), Cuint(dim)))
+    return R
+end
+
+for i = 0:10
+    args = ntuple(x-> Symbol("arg_", x), i)
+    fargs = ntuple(x-> :(broadcast_index($(args[x]), length, global_index)), i)
+    @eval begin
+        # http://developer.amd.com/resources/articles-whitepapers/opencl-optimization-case-study-simple-reductions/
+        function reduce_kernel(state, f, op, v0::T, A, ::Val{LMEM}, result, $(args...)) where {T, LMEM}
+            ui0 = Cuint(0); ui1 = Cuint(1); ui2 = Cuint(2)
+            tmp_local = @LocalMemory(state, T, LMEM)
+            global_index = linear_index(state)
+            acc = v0
+            # # Loop sequentially over chunks of input vector
+            while global_index <= length(A)
+                element = f(A[global_index], $(fargs...))
+                acc = op(acc, element)
+                global_index += global_size(state)
+            end
+            # Perform parallel reduction
+            local_index = threadidx_x(state) - ui1
+            tmp_local[local_index + ui1] = acc
+            synchronize_threads(state)
+
+            offset = blockdim_x(state) ÷ ui2
+            while offset > ui0
+                if (local_index < offset)
+                    other = tmp_local[local_index + offset + ui1]
+                    mine = tmp_local[local_index + ui1]
+                    tmp_local[local_index + ui1] = op(mine, other)
+                end
+                synchronize_threads(state)
+                offset = offset ÷ ui2
+            end
+            if local_index == ui0
+                result[blockidx_x(state)] = tmp_local[ui1]
+            end
+            return
+        end
+    end
+
+end
+
+to_cpu(x) = x
```

Comment:
Do we need this blanket function?

Reply:
Yes, right below?

Reply:
but do we expect this to be called on non-arrays? I have been bitten by things like this before, because it doesn't throw an error if I need to implement it.

Reply:
Well, it's really just supposed to convert `GPUArrays` to `Arrays` and leave anything else alone.
Other checks should be implemented somewhere else!